File: src/main/java/com/illuzionzstudios/mist/ui/render/ItemCreator.java
/**
 * Copyright © 2020 Property of Illuzionz Studios, LLC
 * All rights reserved. No part of this publication may be reproduced, distributed, or
 * transmitted in any form or by any means, including photocopying, recording, or other
 * electronic or mechanical methods, without the prior written permission of the publisher,
 * except in the case of brief quotations embodied in critical reviews and certain other
 * noncommercial uses permitted by copyright law. Any licensing of this software overrides
 * this statement.
 */
package com.illuzionzstudios.mist.ui.render;

import com.illuzionzstudios.mist.Mist;
import com.illuzionzstudios.mist.compatibility.ServerVersion;
import com.illuzionzstudios.mist.compatibility.UItemFlag;
import com.illuzionzstudios.mist.compatibility.UMaterial;
import com.illuzionzstudios.mist.compatibility.UProperty;
import com.illuzionzstudios.mist.util.TextUtil;
import com.illuzionzstudios.mist.util.Valid;
import lombok.Builder;
import lombok.NonNull;
import lombok.Singular;
import net.minecraft.server.v1_15_R1.MinecraftVersion;
import net.minecraft.server.v1_15_R1.Tuple;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.Damageable;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.*;

/**
 * Utility class to easily build custom items. We can set flags,
 * names and lore, and enchantments.
 */
@Builder
public final class ItemCreator {

    /**
     * The actual item stack this represents
     */
    private final ItemStack item;

    /**
     * The {@link UMaterial} of the item
     */
    private final UMaterial material;

    /**
     * The amount of items in the stack
     */
    @Builder.Default
    private final int amount = 1;

    /**
     * Damage of the item for setting custom metadata
     */
    @Builder.Default
    private final int damage = -1;

    /**
     * The display name of the item
     */
    private final String name;

    /**
     * The lore strings to display
     */
    @Singular
    private final List<String> lores;


    /**
     * The enchants applied for the item mapped by level
     */
    @Singular
    private final Map<Enchantment, Integer> enchants;

    /**
     * The item flags
     */
    @Singular
    private List<UItemFlag> flags;

    /**
     * If the {@link ItemStack} has the unbreakable flag
     */
    private boolean unbreakable;

    /**
     * Should we hide all tags from the item (enchants, etc.)?
     */
    @Builder.Default
    private boolean hideTags = false;

    /**
     * Should we add glow to the item? (adds a fake enchant and uses
     * item flags to hide it)
     *
     * The enchant is visible on older MC versions.
     */
    private final boolean glow;

    /**
     * The actual metadata of the item stack
     */
    private final ItemMeta meta;

    //  -------------------------------------------------------------------------
    //  Constructing
    //  -------------------------------------------------------------------------

    /**
     * @return This item suitable for a {@link com.illuzionzstudios.mist.ui.UserInterface}
     */
    public ItemStack makeUIItem() {
        unbreakable = true;
        hideTags = true;

        return make();
    }

    /**
     * Finally construct the {@link ItemStack} from all parameters
     *
     * @return The built item
     */
    public ItemStack make() {

        // Make sure base item and material are set
        Valid.checkBoolean((material != null && material.getMaterial() != null) || item != null, "Material or item must be set!");

        // Actual item we're building on
        ItemStack stack = item != null ? item.clone() : new ItemStack(material.getMaterial(), amount);
        final ItemMeta stackMeta = meta != null ? meta.clone() : stack.getItemMeta();

        Valid.checkNotNull(stackMeta, "Item metadata was somehow null");

        // Skip if trying to build on air
        if (material == UMaterial.AIR)
            return stack;

        // Ensure flags exist
        flags = flags != null ? flags : new ArrayList<>();

        // Set damage
        if (damage != -1) {
            try {
                stack.setDurability((short) damage);
            } catch (final Throwable ignored) {
            }

            try {
                if (stackMeta instanceof Damageable) {
                    ((Damageable) stackMeta).setDamage(damage);
                }
            } catch (final Throwable ignored) {
            }
        }

        // Glow
        if (glow) {
            stackMeta.addEnchant(Enchantment.DURABILITY, 1, true);

            flags.add(UItemFlag.HIDE_ENCHANTS);
        }

        // Enchantments
        if (enchants != null) {
            for (final Enchantment ench : enchants.keySet()) {
                stackMeta.addEnchant(ench, enchants.get(ench), true);
            }
        }

        // Name and lore
        if (name != null) {
            stackMeta.setDisplayName(Mist.colorize("&r" + name));
        }

        if (lores != null && !lores.isEmpty()) {
            final List<String> coloredLores = new ArrayList<>();

            lores.forEach(line -> coloredLores.add(Mist.colorize("&7" + line)));
            stackMeta.setLore(coloredLores);
        }

        // Unbreakable
        if (unbreakable) {
            flags.add(UItemFlag.HIDE_ATTRIBUTES);
            flags.add(UItemFlag.HIDE_UNBREAKABLE);

            if (ServerVersion.olderThan(ServerVersion.V.v1_12)) {
                try {
                    final Object spigot = stackMeta.getClass().getMethod("spigot").invoke(stackMeta);

                    spigot.getClass().getMethod("setUnbreakable", boolean.class).invoke(spigot, true);

                } catch (final Throwable ignored) {
                    // Probably 1.7.10, tough luck
                }
            } else {
                UProperty.UNBREAKABLE.apply(stackMeta, true);
            }
        }

        // Hide flags
        if (hideTags) {
            for (final UItemFlag f : UItemFlag.values()) {
                if (!flags.contains(f)) {
                    flags.add(f);
                }
            }
        }

        // Apply flags
        for (final UItemFlag flag : flags) {
            try {
                stackMeta.addItemFlags(ItemFlag.valueOf(flag.toString()));
            } catch (final Throwable ignored) {
            }
        }

        // Finally apply metadata
        item.setItemMeta(stackMeta);

        return item;
    }

    /**
     * Convenience method to get a new item creator with material, name and lore set
     *
     * @param material The {@link UMaterial} to set
     * @param name The name of the item
     * @param lore Collection of lore strings
     * @return THe builder with these properties
     */
    public static ItemCreatorBuilder of(final UMaterial material, final String name, @NonNull final Collection<String> lore) {
        return of(material, name, lore.toArray(new String[lore.size()]));
    }

    /**
     * See {@link #of(UMaterial, String, Collection)}
     */
    public static ItemCreatorBuilder of(final UMaterial material, final String name, @NonNull final String... lore) {
        return ItemCreator.builder().material(material).name("&r" + name).lores(Arrays.asList(lore)).hideTags(true);
    }

    /**
     * Get a new item creator from material
     *
     * @param mat existing material
     * @return the new item creator
     */
    public static ItemCreatorBuilder of(final UMaterial mat) {
        Valid.checkNotNull(mat, "Material cannot be null!");

        return ItemCreator.builder().material(mat);
    }

}

File: src/main/kotlin/com/illuzionzstudios/mist/item/ItemCreator.kt
package com.illuzionzstudios.mist.item

import com.cryptomorin.xseries.XEnchantment
import com.cryptomorin.xseries.XMaterial
import com.illuzionzstudios.mist.compatibility.ServerVersion
import com.illuzionzstudios.mist.compatibility.ServerVersion.V
import com.illuzionzstudios.mist.compatibility.XItemFlag
import com.illuzionzstudios.mist.compatibility.XProperty
import com.illuzionzstudios.mist.util.TextUtil
import com.illuzionzstudios.mist.util.TextUtil.formatText
import com.illuzionzstudios.mist.util.Valid
import com.illuzionzstudios.mist.util.Valid.checkBoolean
import com.illuzionzstudios.mist.util.Valid.checkNotNull
import lombok.*
import org.bukkit.ChatColor
import org.bukkit.enchantments.Enchantment
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.Damageable
import org.bukkit.inventory.meta.ItemMeta
import java.util.*
import java.util.function.Consumer

/**
 * Utility class to easily build custom items. We can set flags,
 * names and lore, and enchantments. Provides a way to
 * just easily construct an item
 */
@Builder
class ItemCreator {
    /**
     * The actual item stack this represents
     */
    private val item: ItemStack? = null

    /**
     * The [XMaterial] of the item
     */
    private val material: XMaterial? = null

    /**
     * The amount of items in the stack
     */
    @Builder.Default
    private val amount = 1

    /**
     * Damage to the item for setting custom metadata
     */
    @Builder.Default
    private val damage = -1

    /**
     * Custom model data
     */
    @Builder.Default
    private val customModelData = 0

    /**
     * The display name of the item
     */
    private val name: String? = null

    /**
     * The lore strings to display
     */
    @Singular
    private val lores: List<String>? = null

    /**
     * The enchants applied for the item mapped by level
     */
    @Singular
    private val enchants: Map<XEnchantment, Int>? = null

    /**
     * The item flags
     */
    @Singular
    private val flags: MutableList<XItemFlag> = ArrayList()

    /**
     * The actual metadata of the item stack
     */
    private val meta: ItemMeta? = null

    /**
     * If the [ItemStack] has the unbreakable flag
     */
    private var unbreakable = false

    /**
     * Should we hide all tags from the item (enchants, etc.)?
     */
    @Builder.Default
    private var hideTags = false

    /**
     * Should we add glow to the item? (adds a fake enchant and uses
     * item flags to hide it)
     *
     *
     * The enchant is visible on older MC versions.
     */
    @Builder.Default
    private val glow = false

    /**
     * @return This item suitable for a [com.illuzionzstudios.mist.ui.UserInterface]
     */
    fun makeUIItem(): ItemStack {
        unbreakable = true
        hideTags = true
        return make()
    }

    /**
     * Finally construct the [ItemStack] from all parameters
     *
     * @return The built item
     */
    fun make(): ItemStack {
        // Make sure base item and material are set
        Valid.checkBoolean(
            material != null && material.parseMaterial() != null || item != null,
            "Material or item must be set!"
        )

        // Actual item we're building on
        val stack = item?.clone() ?: ItemStack(
            material!!.parseMaterial()!!, amount
        )
        val stackMeta = meta?.clone() ?: stack.itemMeta!!
        Valid.checkNotNull(stackMeta, "Item metadata was somehow null")

        // Skip if trying to build on air
        if (material == XMaterial.AIR) return stack

        // Set damage
        if (damage != -1) {
            try {
                stack.durability = damage.toShort()
            } catch (ignored: Throwable) {
            }
            try {
                if (stackMeta is Damageable) {
                    stackMeta.damage = damage
                }
            } catch (ignored: Throwable) {
            }
        }

        // Custom model data only in 1.14+
        if (ServerVersion.atLeast(V.v1_14) && customModelData != 0) stackMeta.setCustomModelData(customModelData)

        // Glow
        if (glow) {
            stackMeta.addEnchant(Enchantment.DURABILITY, 1, true)
            flags.add(XItemFlag.HIDE_ENCHANTS)
        }

        // Enchantments
        if (enchants != null) {
            for (ench in enchants.keys) {
                stackMeta.addEnchant(ench.parseEnchantment()!!, enchants[ench]!!, true)
            }
        }

        // Name and lore
        if (name != null) {
            stackMeta.setDisplayName(TextUtil.formatText("&r$name"))
        }
        if (lores != null && !lores.isEmpty()) {
            val coloredLores: MutableList<String> = ArrayList()
            lores.forEach(Consumer { line: String ->
                // Colour and split by \n
                val lines = Arrays.asList(*line.split("\\r?\\n".toRegex()).toTypedArray())
                // Append '&7' before every line instead of ugly purple italics
                lines.forEach(Consumer { line2: String -> coloredLores.add(TextUtil.formatText(ChatColor.GRAY.toString() + line2)) })
            })
            stackMeta.lore = coloredLores
        }

        // Unbreakable
        if (unbreakable) {
            if (ServerVersion.olderThan(V.v1_12)) {
                try {
                    val spigot = stackMeta.javaClass.getMethod("spigot").invoke(stackMeta)
                    spigot.javaClass.getMethod("setUnbreakable", Boolean::class.javaPrimitiveType).invoke(spigot, true)
                } catch (ignored: Throwable) {
                    // Probably 1.7.10, tough luck
                }
            } else {
                XProperty.UNBREAKABLE.apply(stackMeta, true)
            }
        }

        // Hide flags
        if (hideTags) {
            for (f in XItemFlag.values()) {
                if (!flags.contains(f)) {
                    flags.add(f)
                }
            }
        }

        // Apply flags
        for (flag in flags) {
            try {
                stackMeta.addItemFlags(ItemFlag.valueOf(flag.toString()))
            } catch (ignored: Throwable) {
            }
        }

        // Finally apply metadata
        stack.itemMeta = stackMeta
        return stack
    }

    companion object {
        //  -------------------------------------------------------------------------
        //  Constructing
        //  -------------------------------------------------------------------------
        /**
         * Convenience method to get a new item creator with material, name and lore set
         *
         * @param material The [XMaterial] to set
         * @param name     The name of the item
         * @param lore     Collection of lore strings
         * @return THe builder with these properties
         */
        fun of(material: XMaterial?, name: String?, lore: Collection<String>): ItemCreator.ItemCreatorBuilder {
            return of(material, name, *lore.toTypedArray())
        }

        /**
         * See [.of]
         */
        fun of(material: XMaterial?, name: String?, vararg lore: String?): ItemCreator.ItemCreatorBuilder {
            return ItemCreator.builder().material(material).name(name).lores(Arrays.asList(*lore)).hideTags(true)
        }

        /**
         * Get a new item creator from material
         *
         * @param mat existing material
         * @return the new item creator
         */
        fun of(mat: XMaterial?): ItemCreator.ItemCreatorBuilder {
            Valid.checkNotNull(mat, "Material cannot be null!")
            return ItemCreator.builder().material(mat)
        }
    }
}